plutus-contract-exe/src/Language/Plutus/Contract/Contract.hs

```diff
@@ -0,0 +1,120 @@
+{-# LANGUAGE FlexibleInstances #-}
+{-# LANGUAGE LambdaCase        #-}
+{-# LANGUAGE TypeFamilies      #-}
+module Language.Plutus.Contract.Contract(
+      Contract
+    , done
+    , emit
+    , waiting
+    , applyInput
+    , applyInputs
+    , drain
+    , outputs
+    , isDone
+    , await
+    , result
+    , loopM
+    ) where
+
+import           Control.Applicative (liftA2)
+import           Control.Monad       ((>=>))
+import           Data.Bifunctor      (first)
+import           Data.List           (foldl')
+
+data Contract t i a =
+    Waiting (i -> Contract t i a)
+    | Done
```

Comment:
Do we need `Done`?

Reply:
Probably not. But we might need an error / `Throw` case at some point